File: live-test/java/src/main/java/demo/MultiSuper.java
package demo;

interface IRender {
    String name();
}

interface IDisplay {
    String name();
}

abstract class Base {
    abstract String name();
}

class Triple extends Base implements IRender, IDisplay {
    @Override
    public String name() {
        return "triple";
    }
}

abstract class ChainBase {
    abstract String tag();
}

class ChainMid1 extends ChainBase {
    @Override
    String tag() {
        return "mid1";
    }
}

interface ITaggable {
    String tag();
}

class ChainMid2 extends ChainMid1 implements ITaggable {
    @Override
    public String tag() {
        return "mid2";
    }
}

class ChainLeaf extends ChainMid2 {
    @Override
    public String tag() {
        return "leaf";
    }
}

interface DiamondTop {
    String pick();
}

interface DiamondLeft extends DiamondTop {
}

interface DiamondRight extends DiamondTop {
}

abstract class DiamondLegacy {
    abstract String pick();
}

class DiamondBottom extends DiamondLegacy implements DiamondLeft, DiamondRight {
    @Override
    public String pick() {
        return "bottom";
    }
}

interface Greeter {
    default String greet() {
        return "hello";
    }
}

interface FriendlyGreeter {
    default String greet() {
        return "hi";
    }
}

abstract class AbstractGreeter {
    abstract String greet();
}

class LoudGreeter extends AbstractGreeter implements Greeter, FriendlyGreeter {
    @Override
    public String greet() {
        return "HELLO";
    }
}

sealed interface SealedShape permits SealedSquare, SealedTriangle {
    int sides();
}

interface Polygon {
    int sides();
}

interface Quadrilateral {
    int sides();
}

final class SealedSquare implements SealedShape, Polygon, Quadrilateral {
    @Override
    public int sides() {
        return 4;
    }
}

final class SealedTriangle implements SealedShape, Polygon {
    @Override
    public int sides() {
        return 3;
    }
}

interface Named {
    String label();
}

interface Tagged {
    String label();
}

interface Identified {
    String label();
}

record LabelPoint(String label, int x, int y) implements Named, Tagged, Identified {
}

interface IntFn {
    int apply(int x);
}

interface IntOp {
    int apply(int x);
}

enum Op implements IntFn, IntOp {
    ADD {
        @Override
        public int apply(int x) {
            return x + 1;
        }
    },
    SUB {
        @Override
        public int apply(int x) {
            return x - 1;
        }
    };

    public abstract int apply(int x);
}

class AnonHost {
    Runnable make() {
        // Anonymous classes cannot have multiple direct supers in Java — kept as 1-super negative anchor
        return new Runnable() {
            @Override
            public void run() {
                System.out.println("anon");
            }
        };
    }
}

File: live-test/kotlin/src/main/kotlin/demo/MultiSuper.kt
package demo

interface IRender {
    fun name(): String
}

interface IDisplay {
    fun name(): String
}

abstract class Base {
    abstract fun name(): String
}

class Triple : Base(), IRender, IDisplay {
    override fun name(): String = "triple"
}

// Deep override chain: Base2 -> Mid1 -> Mid2 -> Leaf.
// Mid2 also implements IComputable so Mid2.m has 2 direct supers (Mid1 + IComputable).
abstract class Base2 {
    abstract fun m(): String
}

interface IComputable {
    fun m(): String
}

open class Mid1 : Base2() {
    override fun m(): String = "mid1"
}

open class Mid2 : Mid1(), IComputable {
    override fun m(): String = "mid2"
}

class Leaf : Mid2() {
    override fun m(): String = "leaf"
}

// Diamond + abstract base: AbstractMethodHolder + Left + Right -> Bottom
abstract class AbstractMethodHolder {
    abstract fun m(): String
}

interface Top {
    fun m(): String
}

interface Left : Top

interface Right : Top

class Bottom : AbstractMethodHolder(), Left, Right {
    override fun m(): String = "bottom"
}

// Default interface method + abstract base + extra interface.
interface Greeter {
    fun greet(): String = "default"
}

interface FriendlyGreeter {
    fun greet(): String = "hi"
}

abstract class AbstractGreeter {
    abstract fun greet(): String
}

class LoudGreeter : AbstractGreeter(), Greeter, FriendlyGreeter {
    override fun greet(): String = "LOUD"
}

// Property override (val) across abstract class + 2 interfaces.
interface Labeled {
    val label: String
}

interface Marked {
    val label: String
}

abstract class AbstractTagged {
    abstract val label: String
}

class Tagged : AbstractTagged(), Labeled, Marked {
    override val label: String = "tag"
}

// Property override (var) across open class + interface.
open class MutableHolder {
    open var holder: String = "init"
}

interface IHolder {
    var holder: String
}

class CustomHolder : MutableHolder(), IHolder {
    override var holder: String = "custom"
        get() = field.uppercase()
        set(value) { field = value.lowercase() }
}

// Sealed class + 2 interfaces declaring the same method.
sealed class Status {
    abstract fun status(): String
}

interface IStatus {
    fun status(): String
}

interface IStateBearer {
    fun status(): String
}

class Active : Status(), IStatus, IStateBearer {
    override fun status(): String = "active"
}

class Inactive : Status(), IStatus, IStateBearer {
    override fun status(): String = "inactive"
}

// Data class implementing 3 interfaces that all declare `name`.
interface Named {
    val name: String
}

interface Identifiable {
    val name: String
}

interface Labelable {
    val name: String
}

data class Point(override val name: String) : Named, Identifiable, Labelable

// Object declaration implementing the same 3 interfaces.
object Singleton : Named, Identifiable, Labelable {
    override val name: String = "singleton"
}
